staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go

```diff
@@ -36,6 +38,125 @@ const (
 	endpoint = "unix:///@kms-socket.sock"
 )
 
+// TestKMSPluginLateStart tests the scenario where kms-plugin pod/container starts after kube-apiserver pod/container.
+// Since the Dial to kms-plugin is non-blocking we expect the construction of gRPC service to succeed even when
+// kms-plugin is not yet up - dialing happens in the background.
+func TestKMSPluginLateStart(t *testing.T) {
+	var callTimeout = 3 * time.Second
+
+	service, err := NewGRPCService(endpoint, callTimeout)
+	if err != nil {
+		t.Fatalf("failed to create envelope service, error: %v", err)
+	}
+	defer destroyService(service)
+
+	time.Sleep(callTimeout / 2)
+	server, err := startTestKMSProvider()
+	if err != nil {
+		t.Fatalf("failed to start test KMS provider server, error: %v", err)
+	}
+	defer stopTestKMSProvider(server)
+
+	data := []byte("test data")
+	_, err = service.Encrypt(data)
+	if err != nil {
+		t.Fatalf("failed when execute encrypt, error: %v", err)
+	}
+}
+
+// TestIntermittentConnectionLoss tests the scenario where the connection with kms-plugin is intermittently lost.
+func TestIntermittentConnectionLoss(t *testing.T) {
+	var (
+		wg1      sync.WaitGroup
+		wg2      sync.WaitGroup
+		timeout  = 10 * time.Second
+		blackOut = timeout / 2
+	)
+	// Start KMS Plugin
+	server, err := startTestKMSProvider()
+	if err != nil {
+		t.Fatalf("failed to start test KMS provider server, error: %v", err)
+	}
+	defer stopTestKMSProvider(server)
```

Comment:
Remove this defer. You have 1 stop below after NewGRPCService and 1 deferred stop after restarting the provider